scripts/lib/CIME/hist_utils.py

```diff
@@ -18,30 +17,33 @@ def _iter_model_file_substrs(case):
     for model in models:
         yield model
 
-def _get_all_hist_files(testcase, model, from_dir, suffix=""):
+def _get_all_hist_files(model, from_dir, suffix="", file_extensions=None, ref_case=None):
     suffix = (".{}".format(suffix) if suffix else "")
 
+    test_hists = []
     # Match hist files produced by run
-    test_hists = glob.glob("{}/{}.{}*.h?.nc{}".format(from_dir, testcase, model, suffix))
-    test_hists.extend(glob.glob("{}/{}.{}*.h.nc{}".format(from_dir, testcase, model, suffix)))
-
-    # Match multi-instance files produced by run
-    test_hists.extend(glob.glob("{}/{}.{}*.h?.*.nc{}".format(from_dir, testcase, model, suffix)))
-    test_hists.extend(glob.glob("{}/{}.{}*.h.*.nc{}".format(from_dir, testcase, model, suffix)))
-
-    # suffix == "" implies baseline comparison, baseline hist files have simpler names
-
-    if suffix == "":
-        test_hists.extend(glob.glob("{}/{}*.h.nc".format(from_dir, model)))
-        test_hists.extend(glob.glob("{}/{}*.h?.nc".format(from_dir, model)))
-        test_hists.extend(glob.glob("{}/{}*.h.*.nc".format(from_dir, model)))
-        test_hists.extend(glob.glob("{}/{}*.h?.*.nc".format(from_dir, model)))
-
+    for extension in file_extensions:
+        if 'initial' in extension:
+            continue
+        if extension.endswith('$'):
+            extension = extension[:-1]
+        string = model+r'\d?_?\d*\.'+extension+suffix+'$'
```

Comment:
This will only work if `<hist_file_extension>` is a regex that goes all the way up through the `.nc`. That is mostly true for CESM, with the exception of:

A few components have this for DART (which according to @jedwards4b may be okay):
```xml
<hist_file_extension>e</hist_file_extension>
```

POP has this; I'm not sure if this one should be captured by this match or not:
```xml
    <hist_file_extension>d[dovt]</hist_file_extension>
```

CISM has this, though in this case it's okay that it isn't picked up (we skip this 'initial_hist' file anyway):
```xml
    <hist_file_extension>initial_hist</hist_file_extension>
```

ESP has this... it may be okay that this isn't picked up:
```xml
      <hist_file_extension>[ei]</hist_file_extension>
```

**However, e3sm has a lot of instances of hist_file_extension not going up through the .nc.**

At the very least, the e3sm config_archive file needs to be changed. Even so, this leaves me a little worried about the asymmetry between how `hist_file_extension` is listed compared with `rest_file_extension`, as well as inconsistencies within `hist_file_extension` itself: We seem to now have a rule that `hist_file_extension` should be a regex going all the way up through the `.nc` _if_ you want these history files to be picked up by the test system. If that is the intended rule, can it be documented somewhere?